core-model/src/main/java/com/sequenceiq/cloudbreak/dto/StackDto.java

```diff
@@ -311,16 +315,29 @@ public Set<Node> getAllFunctioningNodes() {
         Set<Node> ret = new HashSet<>();
         getInstanceGroupDtos().forEach(ig -> {
             InstanceGroupView instanceGroup = ig.getInstanceGroup();
-            ig.getNotDeletedAndNotZombieInstanceMetaData().forEach(im -> {
-                if (StringUtils.isNotBlank(im.getDiscoveryFQDN())) {
-                    ret.add(new Node(im.getPrivateIp(), im.getPublicIp(), im.getInstanceId(),
-                            instanceGroup.getTemplate().getInstanceType(), im.getDiscoveryFQDN(), instanceGroup.getGroupName()));
-                }
-            });
+            List<InstanceMetadataView> notDeletedInstanceMetaData = ig.getNotDeletedInstanceMetaData();
+            returnOnlyLastNodeIfPrivateIpsAreTheSame(notDeletedInstanceMetaData).values().forEach(im -> {
```

Comment:
Maybe rename to filterDuplicatedPrivateIps or something similar? returnOnlyLastNode is a bit misleading.

Reply:
And it is not returning, more likely adding Node-s to a set.